base/shell/explorer/startup.cpp

```diff
@@ -332,3 +384,84 @@ ProcessStartupItems(VOID)
 
     return res ? 0 : 101;
 }
+
+static BOOL IsSelfExplorer(VOID)
+{
+    WCHAR szPath[MAX_PATH];
+
+    GetModuleFileName(NULL, szPath, _countof(szPath));
+    PathRemoveArgsW(szPath);
+    PathRemoveBlanksW(szPath);
+
+    return (StrCmpIW(PathFindFileNameW(szPath), L"explorer.exe") == 0);
+}
+
+BOOL DoFinishStartupItems(VOID)
+{
+    if (s_hStartupMutex)
+    {
+        ReleaseMutex(s_hStartupMutex);
+        CloseHandle(s_hStartupMutex);
+        s_hStartupMutex = NULL;
+    }
+    return TRUE;
+}
+
+BOOL DoStartStartupItems(ITrayWindow *Tray)
+{
+    DWORD dwWait;
+
+    if (!IsSelfExplorer())
```

Comment:
Under what circumstances would this be called from outside explorer?

Reply:
According to my API Monitor investigation, I think Windows Explorer does check its file title (explorer.exe) in its starting up.

Reply:
The user can rename explorer.exe as another name.

Reply:
Shall we delete this self check codes [Y/n]?

Reply:
What if....
..... on Windows, you consistenly rename explorer.exe to e.g. "myshell.exe" (both in SystemRoot\system32\dllcache and in SystemRoot, so that SFC checking doesn't replace explorer.exe back), and set "myshell.exe" in the registry as the new shell to start? I think, that in this situation "myshell.exe" should detect itself as the currrently-running shell.

So... a similar code like this may be kept, BUT the test `return (StrCmpIW(PathFindFileNameW(szPath), L"explorer.exe") == 0);` in the `IsSelfExplorer()` function must definitively be changed to something more robust.

Reply:
As far as I can see, in the WinMain function of explorer.exe, such a code is run:
```
211     if (GetShellWindow() == NULL)
212         CreateShellDesktop = TRUE;
```
Basically, if CreateShellDesktop is TRUE then it means explorer is going to be started as **THE** shell, otherwise it means another shell (whatever it is) is already running, so that this explorer instance should just be considered as folder viewer, not shell.
You can try to leverage the `CreateShellDesktop` into some kind of global boolean value `AmITheShell` (use a different name :) ) that can later be used (e.g. here) to determine whether this instance of explorer has been started as the current shell, or not.

Reply:
![renaming](https://user-images.githubusercontent.com/2107452/41750807-db120492-75f8-11e8-9c5c-d015f9b81158.png)
On 2k3, I copyed explorer.exe and renamed it as another name. It doesn't block opening the folder view.